### PR TITLE
fix: boost score for fully match in matchfuzzy

### LINKED ISSUE #15654: Prefer full match with matchfuzzy
**Is your feature request about something that is currently impossible or hard to do? Please describe the problem.**
Currently `matchfuzzy` functions sorts matches where full match has less score than partial match:

```
echo matchfuzzy(["hello", "lCursor", "Cursor"], "Cursor")
# ['lCursor', 'Cursor'] Cursor is fully matched but lCursor sorted to be first
echo matchfuzzy(["hello", "llCursor", "Cursor"], "Cursor")
# ['llCursor', 'Cursor'] Cursor is fully matched but llCursor sorted to be first
echo matchfuzzy(["hello", "lllCursor", "Cursor"], "Cursor")
# ['Cursor', 'lllCursor'] Cursor is fully matched and first, which is expected

echo matchfuzzy(["hello", "lcursor", "cursor"], "cursor")
# ['cursor', 'lcursor'] here it is fine, expected

```
**Describe the solution you'd like**
Fully matched candidates should be ranked higher than partially matched.

```
echo matchfuzzy(["hello", "lCursor", "llCursor", "nCursor", "rCursor", "Cursor"], "Cursor")
# current result: ['lCursor', 'nCursor', 'rCursor', 'llCursor', 'Cursor']
# expected result ['Cursor', 'lCursor', 'nCursor', 'rCursor', 'llCursor']
```

### PR DESCRIPTION
**Pull request by glepnir**
Problem: Fuzzy matching did not prioritize exact matches and had insufficient handling of multi-byte characters, leading to incorrect scoring.

Solution: Added exact match prioritization with a bonus and improved multi-byte character handling to ensure accurate scoring and ranking.

Fix https://github.com/vim/vim/issues/15654

cc @yegappan

### CODE REVIEW

File: src/testdir/test_matchfuzzy.vim [hunk 1]
```diff
@@ -23,6 +23,8 @@ func Test_matchfuzzy()
   call assert_equal(['aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'], matchfuzzy(['aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'], 'aa'))
   call assert_equal(256, matchfuzzy([repeat('a', 256)], repeat('a', 256))[0]->len())
   call assert_equal([], matchfuzzy([repeat('a', 300)], repeat('a', 257)))
+  " fully match have highest score
```
Comment (zeertzjq) [inline src/testdir/test_matchfuzzy.vim]:
```suggestion
  " full match has highest score
```

File: src/search.c [hunk 2]
```diff
@@ -4452,7 +4453,14 @@ fuzzy_match_compute_score(
 	    // First letter
 	    score += FIRST_LETTER_BONUS;
 	}
+	// Check exact match condition
+        if (currIdx != (int_u)i)
+	    is_exact_match = FALSE;
     }
+    // Boost score for exact matches
+    if (is_exact_match && numMatches == strSz)
```
Comment (chrisbra) [inline src/search.c:4461]: `numMatches == strSz` doesn't seem to make sense? What is the reason for this change please?

Comment (glepnir) [reply src/search.c:4461]: safely check number of matched characters (numMatches) equals the total length of the string

Comment (chrisbra) [reply src/search.c:4461]: Ah, I thought this was the list of candidates.
